File: src/api/routers/stats.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from sqlalchemy import func
from datetime import datetime, timedelta

from ..database import get_db
from ..models import Member, CheckResult, User
from ..dependencies import get_current_active_user

router = APIRouter()

@router.get("/dashboard")
def get_dashboard_stats(db: Session = Depends(get_db), current_user: User = Depends(get_current_active_user)):
    total_members = db.query(Member).count()
    active_members = db.query(Member).filter(Member.membership_status == "active").count()
    lapsed_members = db.query(Member).filter(Member.membership_status == "lapsed").count()
    
    verified = db.query(Member).join(CheckResult).filter(CheckResult.result == "Pass").distinct().count()
    failed = db.query(Member).join(CheckResult).filter(CheckResult.result.in_(["Fail", "Fail_Suburb", "Fail_Street", "Fail_No_Match"])).distinct().count()
    partial = db.query(Member).join(CheckResult).filter(CheckResult.result == "Partial").distinct().count()
    captcha = db.query(Member).join(CheckResult).filter(CheckResult.result == "Captcha").distinct().count()
    
    checked_member_ids = db.query(CheckResult.member_id).distinct().subquery()
    unchecked = db.query(Member).filter(~Member.id.in_(checked_member_ids)).count()
    
    duplicate = db.query(Member).filter(Member.is_duplicate == True).count()
    
    state_distribution = db.query(Member.primary_state, func.count(Member.id)).group_by(Member.primary_state).all()
    
    thirty_days_ago = datetime.utcnow() - timedelta(days=30)
    new_members = db.query(Member).filter(Member.created_at >= thirty_days_ago).count()
    
    return {
        "total_members": total_members,
        "active_members": active_members,
        "lapsed_members": lapsed_members,
        "verified_count": verified,
        "failed_count": failed,
        "partial_match_count": partial,
        "captcha_count": captcha,
        "unchecked_count": unchecked,
        "duplicate_count": duplicate,
        "new_members_30d": new_members,
        "by_state": dict(state_distribution)
    }

@router.get("/electorates")
def get_electorate_stats(db: Session = Depends(get_db), current_user: User = Depends(get_current_active_user)):
    federal_dist = db.query(CheckResult.federal_division, func.count(func.distinct(CheckResult.member_id)))\
        .filter(CheckResult.federal_division.isnot(None))\
        .group_by(CheckResult.federal_division)\
        .order_by(func.count(func.distinct(CheckResult.member_id)).desc()).all()
    
    return [{"federal_division": d, "count": c} for d, c in federal_dist]

File: rust/src/api/routes/stats.rs
use axum::{
    extract::State,
    response::IntoResponse,
    routing::get,
    Json, Router,
};
use serde_json::json;

use crate::api::app::AppState;
use crate::api::auth::AuthUser;

/// GET /stats/dashboard — summary statistics for the dashboard
async fn dashboard_stats(
    State(state): State<AppState>,
    _user: AuthUser,
) -> impl IntoResponse {
    let db = &state.db;

    let total: (i64,) = sqlx::query_as("SELECT COUNT(*) FROM members")
        .fetch_one(db).await.unwrap_or((0,));

    let active: (i64,) = sqlx::query_as(
        "SELECT COUNT(*) FROM members WHERE membership_status = 'active'"
    ).fetch_one(db).await.unwrap_or((0,));

    let lapsed: (i64,) = sqlx::query_as(
        "SELECT COUNT(*) FROM members WHERE membership_status = 'lapsed'"
    ).fetch_one(db).await.unwrap_or((0,));

    let verified: (i64,) = sqlx::query_as(
        "SELECT COUNT(DISTINCT member_id) FROM check_results WHERE result = 'Pass'"
    ).fetch_one(db).await.unwrap_or((0,));

    let failed: (i64,) = sqlx::query_as(
        "SELECT COUNT(DISTINCT member_id) FROM check_results WHERE result IN ('Fail','Fail_Suburb','Fail_Street','Fail_No_Match')"
    ).fetch_one(db).await.unwrap_or((0,));

    let captcha: (i64,) = sqlx::query_as(
        "SELECT COUNT(DISTINCT member_id) FROM check_results WHERE result = 'Captcha'"
    ).fetch_one(db).await.unwrap_or((0,));

    let unchecked: (i64,) = sqlx::query_as(
        "SELECT COUNT(*) FROM members WHERE id NOT IN (SELECT DISTINCT member_id FROM check_results)"
    ).fetch_one(db).await.unwrap_or((0,));

    let thirty_days_ago = chrono::Utc::now() - chrono::Duration::days(30);
    let thirty_days_str = thirty_days_ago.format("%Y-%m-%d %H:%M:%S").to_string();
    let new_30d: (i64,) = sqlx::query_as(
        "SELECT COUNT(*) FROM members WHERE created_at >= ?"
    )
    .bind(&thirty_days_str)
    .fetch_one(db).await.unwrap_or((0,));

    // State distribution
    let state_rows: Vec<(Option<String>, i64)> = sqlx::query_as(
        "SELECT primary_state, COUNT(*) as cnt FROM members GROUP BY primary_state"
    )
    .fetch_all(db).await.unwrap_or_default();

    let by_state: serde_json::Map<String, serde_json::Value> = state_rows
        .into_iter()
        .filter_map(|(st, cnt)| st.map(|s| (s, json!(cnt))))
        .collect();

    Json(json!({
        "total_members": total.0,
        "active_members": active.0,
        "lapsed_members": lapsed.0,
        "verified_count": verified.0,
        "failed_count": failed.0,
        "captcha_count": captcha.0,
        "unchecked_count": unchecked.0,
        "new_members_30d": new_30d.0,
        "by_state": by_state,
    })).into_response()
}

/// GET /stats/electorates — federal division distribution
async fn electorate_stats(
    State(state): State<AppState>,
    _user: AuthUser,
) -> impl IntoResponse {
    let rows: Vec<(Option<String>, i64)> = sqlx::query_as(
        r#"SELECT federal_division, COUNT(DISTINCT member_id) as cnt
           FROM check_results
           WHERE federal_division IS NOT NULL AND federal_division != ''
           GROUP BY federal_division
           ORDER BY cnt DESC"#
    )
    .fetch_all(&state.db)
    .await
    .unwrap_or_default();

    let result: Vec<serde_json::Value> = rows
        .into_iter()
        .map(|(div, cnt)| json!({
            "federal_division": div.unwrap_or_default(),
            "count": cnt
        }))
        .collect();

    Json(result).into_response()
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/dashboard", get(dashboard_stats))
        .route("/electorates", get(electorate_stats))
}
